### ISSUE TITLE
raft::kset syntax

### ISSUE THREAD

**Issue by jonathan-beard** [jonathan-beard/RaftLib#13]
I want to make building complex topologies simpler than it currently is. Specifically, I want to be able to do something like the following:

```cpp
raft::map m;
raft::kernel src,dst;
raft::kernel a, b, c, d;
m += src <= raft::kset( a, b, c, d ) >= dst;
```

This means that there are links as follows:
```
src -> a
src -> b
src -> c
src -> d
a -> dst
b -> dst
c -> dst
d -> dst
```

This syntax seems fairly natural to me, but I wanted to get input from a larger audience before building it into the run-time. I should also note that this syntax addition does not modify any of the current semantics.

-Jonathan

**Comment by pl53** [comment]
How about `m += src <= raft::kset( a, b, c, d ) => dst;`, which looks more symmetric?

**Comment by jonathan-beard** [comment]
Yeah, I like the symmetry...however any thoughts on coding up => as an
overload? That is the real issue with that syntax. The two simpler
candidates are of course: <= >=. One thing I do need to get working
are more complex constructions with the stream overload syntax. You
might have noticed that the <= >= is already present. What's not
perfect is the handling of the more complex topologies. I've gotta
inspect my test case coverage a bit more. I'm certain there are cases
that are missed.

**Comment by adishavit** [comment]
Here are a few thoughts:

1. I think that each line should ideally contain a single operator going in a single direction. I would suggest something like this:
  1. `dst <= raft::kset( a, b, c, d ) <= src` which preserves the order;
  1. `dst << raft::kset( a, b, c, d ) << src` which preserves the order and keeps the streaming concept;
  1. `src >> raft::kset( a, b, c, d ) >> dst` which preserves the example order with the streaming operator;
Having more than one operator is more easy to miss while reading the code.

2. I would suggest considering renaming `kset` to something more semantic like: `dup`, `fork`, `mux`, `demux` etc.

**Comment by jonathan-beard** [comment]
Hmm, I like the renaming idea. I'll comment on the others in a bit, how about "ramify" :) or "diverge" didn't want to use fork since there are too many overloads for the term.

**Comment by adishavit** [comment]
I don't think we need to invent new names.  
What does e.g. [TBB](https://software.intel.com/en-us/node/506211) use?
Broadcast, split, join ?

**Comment by jonathan-beard** [comment]
broadcast might work....

**Comment by adishavit** [comment]
"broadcast" is generic enough to imply both split and join in a single kernel.
BTW, do you expect many such mux-demux (split-then-join) graphs?

**Comment by adishavit** [comment]
or maybe even `tie`? 
Which chimes nicely with the concept of a `tuple` (a k-set). See e.g. [`std::tie`](http://en.cppreference.com/w/cpp/utility/tuple/tie).

**Comment by jonathan-beard** [comment]
I like all those ideas...basically this is a tuple, but we also have to tie
in the concept of kernels and ports. each kernel within the tuple must have
a single port or specified port to connect to.
